src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/CachedDataAnnotationsModelMetadata.cs

```diff
@@ -333,6 +333,31 @@ protected override bool ComputeShowForEdit()
                        : base.ComputeShowForEdit();
         }
 
+        /// <summary>
+        /// Calculate the <see cref="ModelMetadata.TemplateHint"/> value based on presence of a
+        /// <see cref="UIHintAttribute"/> and its <see cref="UIHintAttribute.UIHint"/> value or presence of a
+        /// <see cref="HiddenInputAttribute"/> when no <see cref="UIHintAttribute"/> exists.
+        /// </summary>
+        /// <returns>
+        /// Calculated <see cref="ModelMetadata.TemplateHint"/> value. <see cref="UIHintAttribute.UIHint"/> if a
+        /// <see cref="UIHintAttribute"/> exists. <c>"HiddenInput"</c> if a <see cref="HiddenInputAttribute"/> exists
+        /// and no <see cref="UIHintAttribute"/> exists. <c>null</c> otherwise.
+        /// </returns>
+        protected override string ComputeTemplateHint()
+        {
+            if (PrototypeCache.UIHint != null)
+            {
+                return PrototypeCache.UIHint.UIHint;
+            }
+
+            if (PrototypeCache.HiddenInput != null)
+            {
+                return "HiddenInput";
```

Comment:
:-1: the hint name is only coincidentally the same as the property in the cache.